compiler/runtime/OMRRuntimeAssumptions.cpp

```diff
@@ -40,3 +40,102 @@ void TR::PatchNOPedGuardSite::compensate(bool isSMP, uint8_t *location, uint8_t
    {
    _patchVirtualGuard(location, destination, isSMP);
    }
+
+TR::PatchSites::PatchSites(TR_PersistentMemory *pm, size_t maxSize) :
+    _maxSize(maxSize),
+    _size(0),
+    _refCount(0),
+    _patchPoints(NULL),
+    _firstLocation(NULL),
+    _lastLocation(NULL)
+   {
+   _patchPoints = (uint8_t**) pm->jitPersistentAlloc(sizeof(uint8_t*) * maxSize * 2);
+   }
+
+/**
+ * Add a patch site to the collection.
+ */
+void TR::PatchSites::add(uint8_t *location, uint8_t *destination)
+   {
+   TR_ASSERT_FATAL(_size < _maxSize, "Cannot add more patch sites, max size is %d", _maxSize);
+   _patchPoints[_size * 2] = location;
+   _patchPoints[_size * 2 + 1] = destination;
+   _size++;
+
+   if (_firstLocation == NULL || location < _firstLocation)
+      _firstLocation = location;
+   if (_lastLocation == NULL || location > _lastLocation)
+      _lastLocation = location;
+   }
+
+/**
+ * Compare this collection with another.
+ * Will return true if all locations are the same.
+ */
+bool TR::PatchSites::equals(PatchSites *other)
+   {
+   if (other->getSize() != _size)
+      return false;
+
+   if (_firstLocation != other->getFirstLocation() || _lastLocation != other->getLastLocation())
+      return false;
+
+   for (size_t i = 0; i < _size; ++i)
+      {
+      if (!other->internalContainsLocation(getLocation(i)))
+         return false;
+      }
```

Comment:
This loop only checks to see that the locations in `this->_patchPoints` are a subset of the ones in `other->_patchPoints`. Since the sizes are equal, that would imply equality as long as locations are not allowed to be repeated, which would be reasonable, but maybe it's worth mentioning

Reply:
The original implementation assumed locations would be unique, but the use of this feature in OpenJ9 could not make that guarantee. Moreover, enforcing it could incur significant compile time overhead. Instead, the definition of `equals` has been modified to be true when two lists have the same locations, including duplicates, in the same order.